samples/nrf9160/serial_lte_modem/src/gps/th_gps_service.c

```diff
@@ -0,0 +1,294 @@
+/*
+ * Copyright (c) 2019 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+#define LOG_DOMAIN gps
+#define LOG_LEVEL CONFIG_THIN_LOG_LEVEL
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(LOG_DOMAIN);
+
+#include <zephyr.h>
+#include <stdio.h>
+#include <uart.h>
+#include <string.h>
+
+#include <net/socket.h>
+#include <misc/util.h>
+#include <gps.h>
+#include <lte_lc.h>
+#include "th_connect.h"
+#include "th_util.h"
+
+#if !defined(CONFIG_LTE_NETWORK_MODE_NBIOT_GPS) && \
+	!defined(CONFIG_LTE_NETWORK_MODE_LTE_M_GPS)
+#error Neither LTE_NETWORK_MODE_NBIOT_GPS nor \
+	LTE_NETWORK_MODE_LTE_M_GPS is defined
+#endif
+
+static struct gps_data gps_data;
+static struct k_work send_gps_data_work;
+
+/* START: code adapted from applications\asset_tracker\src\gps_controller */
+
+/* Structure to hold GPS work information */
+static struct {
+	enum {
+		GPS_WORK_START,
+		GPS_WORK_STOP
+	} type;
+	struct k_delayed_work work;
+	struct device *dev;
+} gps_work;
+
+static atomic_t gps_is_active;
+static atomic_t gps_is_enabled;
+
+/* Forward declaration */
+int gps_control_init(gps_trigger_handler_t handler);
+void gps_control_on_trigger(void);
+void gps_control_stop(u32_t delay_ms);
+void gps_control_start(u32_t delay_ms);
+bool gps_control_is_active(void);
+bool gps_control_is_enabled(void);
+void gps_control_enable(void);
+void gps_control_disable(void);
+
+static int start(void)
+{
+	int err;
+
+	err = gps_start(gps_work.dev);
+	if (err) {
+		LOG_ERR("Failed starting GPS!");
+		return err;
+	}
+
+	atomic_set(&gps_is_active, 1);
+
+	LOG_INF("GPS started successfully");
+	LOG_DBG("Searching for satellites to get position fix, ");
+	LOG_DBG("This may take several minutes.");
+	LOG_DBG("The device will attempt to get a fix for %d seconds, ",
+		CONFIG_GPS_CONTROL_FIX_TRY_TIME);
+	LOG_DBG("before the GPS is stopped.");
+
+	return 0;
+}
+
+static int stop(void)
+{
+	int err;
+
+	err = gps_stop(gps_work.dev);
+	if (err) {
+		return err;
+	}
+
+	LOG_INF("GPS stopped successfully");
+	return 0;
+}
+
+static void gps_work_handler(struct k_work *work)
+{
+	int err;
+
+	if (gps_work.type == GPS_WORK_START) {
+		err = start();
+		if (err) {
+			LOG_ERR("GPS could not be started, error: %d", err);
+			return;
+		}
+
+		LOG_INF("GPS operation started");
+
+		atomic_set(&gps_is_active, 1);
+
+		gps_work.type = GPS_WORK_STOP;
+
+		k_delayed_work_submit(&gps_work.work,
+				K_SECONDS(CONFIG_GPS_CONTROL_FIX_TRY_TIME));
+
+		return;
+	} else if (gps_work.type == GPS_WORK_STOP) {
+		err = stop();
+		if (err) {
+			LOG_ERR("GPS could not be stopped, error: %d", err);
+			return;
+		}
+
+		LOG_INF("GPS operation was stopped");
+
+		atomic_set(&gps_is_active, 0);
+
+		if (atomic_get(&gps_is_enabled) == 0) {
```

Comment:
Hmm there seems to be `gps_control_is_enabled` API for that, which is not used otherwise.

Reply:
Let's use the atomic_get internal flag here. The _gps_control_is enabled_ and _gps_control_is_active_ are now exposed to client side. Code is originated from the assert_tracker application, not optimized.